src/auth/AclProxyAuth.cc

```diff
@@ -182,9 +174,23 @@ ACLProxyAuth::matchProxyAuth(ACLChecklist *cl)
             return 0;
         }
     }
+
     /* check to see if we have matched the user-acl before */
-    int result = cacheMatchAcl(&checklist->auth_user_request->user()->proxy_match_cache, checklist);
-    checklist->auth_user_request = NULL;
+    auto &cache = checklist->auth_user_request->user()->proxyAuthAclCache;
+
+    SBuf key;
+    key.appendf("ACL %s(%" PRIX64 ")", name, this);
```

Comment:
This cache key computation raises several red flags for me, but I will start with just one basic question to make sure that I am not missing something important. Please do _not_ adjust the code based on this question (yet).

I assume that the `name` here is the ACL name/label/ID provided by the admin in squid.conf (e.g., `x` in `acl x has request`). I also assume that `this` pointer points to the corresponding ACL object. If my assumptions are correct, why are we using _both_ the name and the object pointer that stores that name as the cache key?

Reply:
You are correct. Name is for readability in the logs where we display the key. "this=" because it was the old key and I have not yet tested whether it can reliably be removed. I am inclined to remove the pointer value if we can because it is a potential security info leak in cache.log.

Reply:
Assuming those are the only two reasonable choices, please change the key to be either the ACL's object name or the ACL object address.

Needless to say, either design would require other adjustments. For example, if we make the key to be the name, then we should probably convert the ACL::name type to SBuf, and if we make the key the pointer, then ClpMap would have to be enhanced to support non-SBuf keys. Using the ACL object address would optimize caching and lookup speed, but other reasons may make SBuf-based caching more attractive. Please try to justify your decision.

> ... the pointer value ... is a potential security info leak in cache.log

IMHO, we should not ban logging of object addresses. Since logging pointers is often useful for debugging, and properly obfuscating logged raw pointers costs too many CPU cycles, we should advise against making cache.log readable to potential adversaries instead.

----

P.S. If you would like to replace this cache with an authentication helper cache instead (as you mentioned [earlier](https://github.com/squid-cache/squid/pull/697#issuecomment-664381601)), I would support that alternative as well (in principle). Or we could add TTL configuration to proxy_auth. Unfortunately, I do not know enough about this code area to warn you of the potential pitfalls ahead of time.

Reply:
Done key update in 940b501. Looking into the ACL::name type change now.